metadata-ingestion/src/datahub/emitter/openapi_emitter.py

```diff
@@ -0,0 +1,188 @@
+from datahub.emitter.mcp import MetadataChangeProposalWrapper
+from datahub.emitter.openapi_tracer import OpenAPITrace
+from datahub.emitter.response_helper import extract_trace_data
+from datahub.emitter.rest_emitter import (
+    _DATAHUB_EMITTER_TRACE,
+    BATCH_INGEST_MAX_PAYLOAD_LENGTH,
+    INGEST_MAX_PAYLOAD_BYTES,
+    DataHubRestEmitter,
+)
+from datahub.emitter.serialization_helper import pre_json_transform
+from datahub.metadata.com.linkedin.pegasus2avro.mxe import (
+    MetadataChangeProposal,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass
+class Chunk:
+    items: List[str]
+    total_bytes: int = 0
+
+    def add_item(self, item: str) -> bool:
+        item_bytes = len(item.encode())
+        if not self.items:  # Always add at least one item even if over byte limit
+            self.items.append(item)
+            self.total_bytes += item_bytes
+            return True
+        self.items.append(item)
+        self.total_bytes += item_bytes
+        return True
+
+    @staticmethod
+    def join(chunk: "Chunk") -> str:
+        return "[" + ",".join(chunk.items) + "]"
+
+
+class DataHubOpenApiEmitter(DataHubRestEmitter, OpenAPITrace):
+    def __init__(
+        self,
+        gms_server: str,
+        token: Optional[str] = None,
+        timeout_sec: Optional[float] = None,
+        connect_timeout_sec: Optional[float] = None,
+        read_timeout_sec: Optional[float] = None,
+        retry_status_codes: Optional[List[int]] = None,
+        retry_methods: Optional[List[str]] = None,
+        retry_max_times: Optional[int] = None,
+        extra_headers: Optional[Dict[str, str]] = None,
+        ca_certificate_path: Optional[str] = None,
+        client_certificate_path: Optional[str] = None,
+        disable_ssl_verification: bool = False,
+        default_trace_mode: bool = False,
+    ):
+        super().__init__(
+            gms_server,
+            token,
+            timeout_sec,
+            connect_timeout_sec,
+            read_timeout_sec,
+            retry_status_codes,
+            retry_methods,
+            retry_max_times,
+            extra_headers,
+            ca_certificate_path,
+            client_certificate_path,
+            disable_ssl_verification,
+            default_trace_mode,
+        )
+
+    def _emit_generic(self, url: str, payload: dict) -> Response:
+        return super()._emit_generic(url, payload=json.dumps(payload))
+
+    def _to_request(
+        self,
+        mcp: Union[MetadataChangeProposal, MetadataChangeProposalWrapper],
+        async_flag: Optional[bool] = None,
+        async_default: bool = False,
+    ):
+        resolved_async_flag = async_flag if async_flag is not None else async_default
+        url = f"{self._gms_server}/openapi/v3/entity/{mcp.entityType}?async={'true' if resolved_async_flag else 'false'}"
+        ensure_has_system_metadata(mcp)
+        aspect_value = pre_json_transform(mcp.aspect.to_obj())
```

Comment:
does this encode stuff properly? I thought that openapi encodes aspects slightly differently than the restli endpoint

Reply:
Seems to work as expected. I wrote this in order to ingest search result entities with all aspects from one instance to another. I've tested it with around 10k random entities. Including schema metadata and lots of different aspects.